### ISSUE TITLE
big gaps in the sequence of session index inside table com_snowplowanalytics_snowplow_client_session_1

### ISSUE THREAD

**Issue by renatoac** [snowplow/snowplow-android-tracker#184]
there are many rows where some user_id's go from session_index with a low number like 1 to a high number like 5000 and there isn't any other session_index inside this range.

**Comment by jbeemster** [comment]
Hi @renatoac would you be able to provide a bit more context around the rows where this occurs.

Are there any common Android Versions and/or Device types where this appears to happen?
What are your current session timeout values?
What version of the Tracker are you running?

The session index only increments when:

* Either the Foreground or Background timeout are triggered.
* The application is closed and re-opened.

So it is very peculiar that it is incrementing so suddenly!

**Comment by renatoac** [comment]
@jbeemster,

We are using the default configs for Foreground and Background timeout.
Here is a case of one user_id: https://gist.github.com/renatoac/9249f9182a19cfad4a58.

For example, see the gap in session_index between lines 500 and 501.

**Comment by renatoac** [comment]
And in the following lines - https://gist.github.com/renatoac/9249f9182a19cfad4a58#file-snowplow-android-tracker-issues-184-csv-L1317-L1318 - there is a odd behavior:  session_index is greater than previous row but tstamp is earlier.

**Comment by bernardosrulzon** [comment]
Hey @jbeemster 

Following up on this, I have two additional datasets that show this behaviour. I'm seeing this on a considerable number of users, and have checked the bad bucket for missing events with no luck.

https://gist.github.com/bernardosrulzon/af55338de4d9c70f62cd

Please let us know if we can provide any other info!

**Comment by jbeemster** [comment]
Hi @renatoac and @bernardosrulzon a quick thought on this.  The gaps appear to occur between large date discrepancies.  In that first case 8 days and 12 hours passed between a session context being sent.

In fact all of the session_count jumps seem to correspond with some form of date jump.  Is it possible the application was backgrounded for these periods which was resulting in the session timing out and incrementing?

Some extra information that could help me figure this out:

1. What version of the Android Tracker are using (Classic or RxJava)?
2. The code used to instantiate the Tracker.
3. Whether the Tracker is being constructed as a singleton object?
4. Are you signalling to the Tracker whether or not the application has gone to the background?
  - [Setup guide](https://github.com/snowplow/snowplow/wiki/Android-Tracker#12-client-sessions)

Some thoughts on avoiding this:

1. We could start automatically sending an `end_of_session` event to capture every time the session expires and then a `start_of_session` event to capture the start.  This might also make it easier to understand when in a session users started doing things.
2. Potentially extending the session timeouts?

@alexanderdean thoughts on this?

**Comment by alexanderdean** [comment]
I like the idea of `end_` and `start_` of `client_session` events...

My question back to @jbeemster is - why would backgrounding for an excessive period lead to *gaps* in the session indexes?

**Comment by jbeemster** [comment]
Unless events are being sent there is nowhere for the context to be attached, as such the count increments steadily but no session context is sent to the Collector, hence you have gaps in the session count.

My idea of start and end would ensure that each session is accounted for, irrespective of whether or not an event occurred during that session.

**Comment by alexanderdean** [comment]
Hmm. I see what you mean. I think I would look at it differently: I don't believe a "session" which wholly takes place in the background is a session. In other words, I don't think we should "keep counting" while an app is backgrounded...

**Comment by jbeemster** [comment]
I see what you mean, it should expire once in the background and then halt.  Only resumed on coming back to foreground.  This however is wholly dependant on the developer signalling to the Tracker that it has indeed gone to the background...

Another way of looking at it could be that we stop counting whenever we have a session that expired with 0 events being sent in it?

**Comment by alexanderdean** [comment]
But don't we ask the Android developer to actively flag a backgrounding or foregrounding event? So if they do that, we should be able to prevent the behavior in this ticket...

**Comment by jbeemster** [comment]
We do yes, I am suggesting doing both so we have a fail-safe for if that implementation is not done.

But in the case of Foreground, if a session expires where no events occurred should we continue to count?

**Comment by alexanderdean** [comment]
I think the trouble with the fail-safe is that you are trying to second-guess the user's behavior in the app based on behaviors in the tracker (no events means no session)... I think we should start with a fix first for session index incrementing in the case of a backgrounded app, and have a separate ticket to explore your failsafe idea further...

**Comment by jbeemster** [comment]
As a quick fix until we get 0.6.0 out developers can also simply shutdown the session checker on going to background using these functions: https://github.com/snowplow/snowplow-android-tracker/blob/master/snowplow-android-core/src/main/java/com/snowplowanalytics/snowplow/tracker/Tracker.java#L480-L489

You will need to resume again on application coming to foreground.

**Comment by alexanderdean** [comment]
Good point!

**Comment by bernardosrulzon** [comment]
@jbeemster @alexanderdean Interesting discussion! Thanks for narrowing down this issue. A few thoughts:
1. We completely agree that a session without events is not a session and should not cause sessionIndex to increment
2. We were able to reproduce this behaviour - when closing opening the app again, sessionIndex is incremented by 2. Raw collector data available [here](https://gist.github.com/bernardosrulzon/4d69961ae6807b706eff)
3. We're also seeing some sessionIndex increases before the minimal timeout. Guess that happens becuase of the previous point, right? Is this expected behaviour? It would make more sense to continue in the same session.

We're using the classic tracker, not implemented as a singleton, and we've setup the `setIsBackground` methods. Not sure I understood your workaround - is that what you mean?

```java
@Override
protected void onPause() {
    super.onPause();
    tracker.getSession().setIsBackground(true);
    tracker.pauseSessionChecking();
}
```

**Comment by jbeemster** [comment]
Yes that is the workaround!  Just make to add `tracker.resumeSessionChecking()` in your `onResume()` function.

As for the need for a singleton, every time a Tracker is created it starts a new service to monitor the session timeouts.  However each service shares the same file, hence you can then get into a situation where:

1. You sent 20 events, so you now have 20 session checkers.
2. The timeout occurs and then the sessionCount is incremented 20 times.

This is because they write to the same file.  The sending is a little bit more forgiving but would also struggle a great deal as you would have 20 emitters all reading from the same event database.

---

I imagine a great deal of these issues will go away when you have a singleton pattern for the Tracker and with that fix in place.

**Comment by bernardosrulzon** [comment]
Ah, that makes sense! Do you have any examples of how to implement the tracker as a singleton?

cc/ @mesquitaa

**Comment by jbeemster** [comment]
So what I generally do for this is:

1. Make a static class.
2. Make a getter function which fetches the static Tracker object.
3. Make static track functions which use said getter function.

__Note__: This example shows the getter function without any null checks.  In production you should really have a null check there.

```java
public class Snowplow {

    private static final String NAMESPACE = "YOUR_NAMESPACE_HERE";
    private static final String APP_ID = "YOUR_APP_ID_HERE";
    private static final String TRACKER_URI = "YOUR_URI_HERE";
    private static Tracker mTracker;

    public static void setup(Context context) {
        Subject mSubject = new Subject.SubjectBuilder()
                .context(context)
                .build();

        Emitter mEmitter = new Emitter.EmitterBuilder(TRACKER_URI, context)
                .method(HttpMethod.POST)
                .option(BufferOption.HeavyGroup)
                .build();

        mTracker = new Tracker.TrackerBuilder(mEmitter, NAMESPACE, APP_ID, context)
                .platform(DevicePlatforms.Mobile)
                .subject(mSubject)
                .threadCount(20)
                .sessionContext(true)
                .level(LogLevel.OFF)
                .build();
    }

    /**
     * @return the Tracker object
     */
    public static Tracker getTracker() {
        return this.mTracker;
    }

    /**
     * Sends an event on Application Boot
     */
    public static void trackBoot() {
        getTracker().track(Structured.builder()
                .action("Boot")
                .property("true")
                .build());
    }
}
```

You can then reference this class from anywhere in your application code by calling:

`Tracker staticTracker = Snowplow.getTracker();`

If you need more code samples please do let me know!

**Comment by jbeemster** [comment]
As a side note @bernardosrulzon this is a known issue and in the next release I plan to refactor the Tracker in such a way as to make the Tracker a singleton by default.  Without any way of making multiple Trackers.

The issues you are seeing in ObjectiveC are likely also caused by the same implementation issue.

**Comment by bernardosrulzon** [comment]
Thanks! Just tested your workaround, but it has the side-effect of not creating a new session after a user spent significant time in background. I guess it's better to leave that unchanged until 0.6.0 comes around.

**Comment by bernardosrulzon** [comment]
@jbeemster It turned out we had already been using a singleton pattern, and we decided not to use your workaround since it basically ignores the time users spend on background. It might be better to have a time-based definition of sessions (on our end) until version 0.6.0 is out.

Is that in your short term pipeline?

Thanks!

**Comment by jbeemster** [comment]
Hey @bernardosrulzon do you have any feedback on this ticket after the 0.6.0 release?  Are you still seeing these types of issues or can we close this off.

**Comment by bernardosrulzon** [comment]
Hey @jbeemster! I'm still seeing these issues in v0.6.1. Will investigate further and let you know.

For the record, this is the base query I'm using to debug:
```
SELECT e.v_tracker,
       app_id,
       cs.user_id,
       session_index
FROM com_snowplowanalytics_snowplow_client_session_1 cs
INNER JOIN atomic.events e ON e.event_id = cs.root_id AND e.collector_tstamp = cs.root_tstamp
WHERE v_tracker = 'andr-0.6.1' AND collector_tstamp > '2017-01-01' and root_tstamp > '2017-01-01'
GROUP BY 1,2,3,4
ORDER BY 3,4;
```

**Comment by bernardosrulzon** [comment]
This problem is now affecting 50% of our sessions. See below the histogram of the # of sessions being "skipped"
![image](https://cloud.githubusercontent.com/assets/6955526/21957446/4c3a28c0-da7d-11e6-8ca7-4b6247d642f2.png)

It actually seems to have gotten worse with 0.6.2 - in v0.5.4 we saw it affect 35% of sessions.

This is the revised query I'm using to debug. Perhaps you could run this with some of your clients to compare?

```
WITH main AS 
(SELECT MIN(derived_tstamp) AS min_derived_tstamp,
        app_id,
        cs.user_id,
        session_index
FROM com_snowplowanalytics_snowplow_client_session_1 cs
INNER JOIN atomic.events e ON e.event_id = cs.root_id AND e.collector_tstamp = cs.root_tstamp
WHERE v_tracker = 'andr-0.6.2' AND collector_tstamp > '2017-01-01' AND root_tstamp > '2017-01-01'
GROUP BY 2,3,4
ORDER BY 3,4)

SELECT *, 
       LAG(session_index) OVER (PARTITION BY user_id ORDER BY session_index) AS previous_session_index,
       LAG(min_derived_tstamp) OVER (PARTITION BY user_id ORDER BY session_index) AS previous_min_derived_tstamp
FROM main;
```

**Comment by alexanderdean** [comment]
Scheduling so this gets investigated...

**Comment by AlexBenny** [comment]
I think this can be definitely fixed with the issue #366
